llvm/unittests/Support/KnownBitsTest.cpp

```diff
@@ -516,6 +532,19 @@ TEST(KnownBitsTest, BinaryExhaustive) {
         return N1.lshr(N2);
       },
       checkOptimalityBinary, /* RefinePoisonToZero */ true);
+  testBinaryOpExhaustive(
+      [](const KnownBits &Known1, const KnownBits &Known2) {
+        return KnownBits::lshr(Known1, Known2, /*ShAmtNonZero=*/false,
+                               /*Exact=*/true);
+      },
+      [](const APInt &N1, const APInt &N2) -> std::optional<APInt> {
+        if (N2.uge(N2.getBitWidth()))
+          return std::nullopt;
+        if (!N2.isZero() && !N1.extractBits(N2.getZExtValue(), 0).isZero())
```

Comment:
This should still work even without the `!N2.isZero()` check.

Reply:
`extractBits` has:
```
  assert(bitPosition < BitWidth && (numBits + bitPosition) <= BitWidth &&
         "Illegal bit extraction");
```

so need `N2 > 0`.

Reply:
We should relax that assertion, I think. It predates zero-length APInts being made legal.

Reply:
@goldsteinn huh? How does that assertion prevent calling it with `numBits == 0`?

Reply:
(Though the assertion probably should be tweaked to `bitPosition **<=** BitWidth && (numBits + bitPosition) <= BitWidth`, to allow for extracting the zero high-order bits.)

Reply:
Oh yeah... I misread `BitWidth` as `numBits`...., ill update.

Reply:
Although `APInt` is somewhat littered with things like `assert(BitWidth && "zero width values not allowed");

The usage we have here should be fine, though.